### ISSUE TITLE
LINQ expressions being evaluated locally when .Include() is involved

### ISSUE THREAD

**Issue by vaindil** [aspnet/EntityFramework#7164]
### Problem

I'm querying a DB using a DbContext I wrote, and I recently added some `.Include()` calls to an existing search function. Since I've added those, I'm getting this warning (with `OrderBy()`, `Skip()`, and `Take()`):

> The LINQ expression 'OrderBy(__p_2)' could not be translated and will be evaluated locally. To configure this warning use the DbContextOptionsBuilder.ConfigureWarnings API (event id 'RelationalEventId.QueryClientEvaluationWarning'). ConfigureWarnings can be used when overriding the DbContext.OnConfiguring method or using AddDbContext on the application service provider.

If I remove the `.Include()`/`.ThenInclude()` expressions, the query translates and no warning is thrown. The problem happens even if only `.Include()` is involved; removing only `.ThenInclude()` does not fix it. Is this expected behavior?

### Code

Query:

```csharp
var query = _context.Posts.AsQueryable();

// various search calls such as this
if (search.UserId.HasValue)
    query = query.Where(p => p.UserId == search.UserId.Value);

// and finally
query = query
    .Include(p => p.User)
        .ThenInclude(u => u.Avatar)
    .OrderBy(p => p.CreatedAt)
    .Skip(start - 1)
    .Take(limit);

return await query.ToListAsync();
```

Base classes:

```csharp
public class User
{
    public long Id { get; set; }
    public string Email { get; set; }
    public string PasswordHash { get; set; }

    // various other basic properties

    public Avatar Avatar { get; set; }
    public ICollection<Post> CreatedPosts { get; set; }
}

public class Avatar
{
    public long UserId { get; set; }
    public string Key { get; set; }

    public User User { get; set; }
}

public class Post
{
    public long Id { get; set; }
    public string Title { get; set; }
    public string Content { get; set; }
    public long UserId { get; set; }
    
    public User User { get; set; }
}
```

I use an extension method for context configuration from [this answer](http://stackoverflow.com/a/35460956/1672458) on Stack Overflow.

```csharp
    static class ModelBuilderExtensions
    {
        public static void AddConfiguration<TEntity>(
            this ModelBuilder modelBuilder,
            DbEntityConfiguration<TEntity> entityConfiguration) where TEntity : class
        {
            modelBuilder.Entity<TEntity>(entityConfiguration.Configure);
        }
    }

    abstract class DbEntityConfiguration<TEntity> where TEntity : class
    {
        public abstract void Configure(EntityTypeBuilder<TEntity> entity);
    }
```

`OnModelCreating`:

```csharp
protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    base.OnModelCreating(modelBuilder);

    modelBuilder.AddConfiguration(new UserConfig());
    modelBuilder.AddConfiguration(new AvatarConfig());
    modelBuilder.AddConfiguration(new PostConfig());
```

Config classes:

```csharp
class UserConfig : DbEntityConfiguration<User>
{
    public override void Configure(EntityTypeBuilder<User> entity)
    {
        entity.ToTable("User");
        entity.HasKey(u => u.Id);

        entity.Property(u => u.Email).IsRequired().HasMaxLength(320);
        entity.Property(u => u.PasswordHash).IsRequired().HasMaxLength(150);

        entity.HasOne(u => u.Avatar)
            .WithOne(a => a.User)
            .HasForeignKey<Avatar>(a => a.UserId);

        entity.HasMany(u => u.CreatedPosts)
            .WithOne(cp => cp.User)
            .HasForeignKey(cp => cp.UserId);
    }
}

class AvatarConfig : DbEntityConfiguration<Avatar>
{
    public override void Configure(EntityTypeBuilder<Avatar> entity)
    {
        entity.ToTable("Avatar");
        entity.HasKey(i => i.UserId);

        entity.Property(pi => pi.Key).IsRequired();
    }
}

class PostConfig : DbEntityConfiguration<Post>
{
    public override void Configure(EntityTypeBuilder<Post> entity)
    {
        entity.ToTable("Post");
        entity.HasKey(p => p.Id);

        entity.Property(p => p.Title).IsRequired().HasMaxLength(150);
        entity.Property(p => p.Content).HasMaxLength(4000);
    }
}
```

### Further technical details

**EF Core version:** 1.1.0
**Operating system:** Windows 10 Education x64
**Visual Studio version:** VS 2015 Enterprise

**Comment by vaindil** [comment]
I haven't been able to work around this. The only thing I can do is remove `.Include()` calls and manually query the DB for each filtered object. Querying the DB 10 separate times is better than loading thousands of entries into memory.

**Comment by maumar** [comment]
Can you also provide code listing for entities (Post, User, Avatar), as well as contents of OnModelCreating method on your DbContext? Another thing to try could be to perform Includes before the filtering.

**Comment by vaindil** [comment]
@maumar I've updated the first comment with that info, sorry!

**Comment by vaindil** [comment]
Any chance of an update on this?

**Comment by maumar** [comment]
@vaindil not yet. We have been busy with the 1.1.1 release of EF Core as well as new version of EF6. Those two are almost done so you can expect progress on new issues shortly. Sorry for the wait!

**Comment by vaindil** [comment]
@maumar Not a problem, just wanted to check in. Thank you!

**Comment by maumar** [comment]
Fixed in f022d66bc3a461851a6250536399cd3b8f8bf8d2

The query we produce now:

```sql
SELECT [p].[Id], [p].[Content], [p].[CreatedAt], [p].[Title], [p].[UserId], [p.User].[Id], [p.User].[Email], [p.User].[PasswordHash], [p.User.Avatar].[UserId], [p.User.Avatar].[Key]
FROM [Post] AS [p]
INNER JOIN [User] AS [p.User] ON [p].[UserId] = [p.User].[Id]
LEFT JOIN [Avatar] AS [p.User.Avatar] ON [p.User].[Id] = [p.User.Avatar].[UserId]
WHERE [p].[UserId] = @__userId_0
ORDER BY [p].[CreatedAt]
OFFSET @__p_1 ROWS FETCH NEXT @__p_2 ROWS ONLY
```
